[Tests] Add options for test to select clouds

Michaelvll opened the pull request:
<!-- Describe the changes in this PR -->

1. Add options to test for easier to select clouds to be used for testing
1. Add options to the `pyproject.toml`, so no need for `run_smoke_tests.sh`, but only `pytest tests/test_smoke.py`
1. Do not error out when downing an Azure cluster whose ResourceGroup not exist.
2. Added a `sky status cluster1 cluster2` option, so the user can filter out the clusters or only refresh the status for the provided cluster names
3. Fix the issue where the `lsof` is not installed in the image.

<!-- Describe the tests ran -->
<!-- Unit tests (tests/test_*.py) are part of GitHub CI; below are tests that launch on the cloud. -->

Tested (run the relevant ones):

- [x] `sky launch --cloud aws -t t3.medium "sudo apt install tree -y"`
- [x] All smoke tests: `pytest tests/test_smoke.py` (`pytest --lf` when there are several failed tests)

concretevitamin commented on sky/cli.py:
```diff
@@ -1359,8 +1359,13 @@ def exec(
-def status(all: bool, refresh: bool):  # pylint: disable=redefined-builtin
+def status(all: bool, refresh: bool, clusters: List[str]):  # pylint: disable=redefined-builtin
     # NOTE(dev): Keep the docstring consistent between the Python API and CLI.
     """Show clusters.
```
```suggestion
    """Show clusters.
    
    If CLUSTERS is given, show those clusters.
```

Michaelvll replied:
Done. Thanks!

Michaelvll replied:
Done. Thanks!

concretevitamin commented on sky/cli.py:
```diff
@@ -1359,8 +1359,13 @@ def exec(
 @usage_lib.entrypoint
-def status(all: bool, refresh: bool):  # pylint: disable=redefined-builtin
+def status(all: bool, refresh: bool, clusters: List[str]):  # pylint: disable=redefined-builtin
     # NOTE(dev): Keep the docstring consistent between the Python API and CLI.
```
Should we keep the interface / docstr consistent?

concretevitamin commented on sky/resources.py:
```diff
@@ -85,9 +85,14 @@ def __init__(
+            else:
+                self._image_id = {
+                    k.strip(): v.strip() for k, v in image_id.items()
+                }
```
Why do these image_ids need strip()?

Michaelvll replied:
It seems when we add a space to the image_id, the launching will fail and can be quite confusing.

concretevitamin commented on sky/clouds/service_catalog/gcp_catalog.py:
```diff
@@ -460,6 +460,7 @@ def get_image_id_from_tag(tag: str, region: Optional[str]) -> Optional[str]:
+def is_image_tag_valid(tag: str, region: Optional[str]) -> bool:
     """Returns whether the image tag is valid."""
-    return common.is_image_tag_valid_impl(_image_df, tag, region)
+    del region  # Unused.
```
Why unused? Thought we could pass it along?

Michaelvll replied:
Our catalog for GCP images do not contain `Region` column, we can either handle it here or in the underlying function. I think it might be ok to do it here, as it is GCP specific. Wdyt?

concretevitamin commented on sky/templates/aws-ray.yml.j2:
```diff
@@ -118,12 +118,15 @@ setup_commands:
   # Line 'mkdir -p ..': disable host key check
   # Line 'python3 -c ..': patch the buggy ray files and enable `-o allow_other` option for `goofys`
-  - sudo systemctl stop unattended-upgrades || true;
+  - function mylsof { p=$(for pid in /proc/{0..9}*; do i=$(basename "$pid"); for file in "$pid"/fd/*; do link=$(readlink -e "$file"); if [ "$link" = "$1" ]; then echo "$i"; fi; done; done); echo "$p"; };
```
Does this reliably handle the apt problem?

Michaelvll replied:
The `mylsof filename` works the same as `lsof filename` returning the pids that hold the file handle, according to the test on a newly launched GCP VM (`lsof /dev/null` vs `mylsof /dev/null`).

concretevitamin reviewed (commented):
Thanks @Michaelvll! This is awesome. Read all files except test_smoke.

Some questions:

1. I'm seeing some randomly generated names, why?
```
Terminating cluster testpiaei-itzn-2b289124-20...done.
...
Terminating cluster testpsaejbmna-etr-2b289124-4f...done.
```

2. `pytest tests/test_smoke.py -v --aws` was expecting it to run AWS-only tests, but seeing
```
...
[azure_region] Test started. Log: less /var/folders/8f/56gzvwkd3n3293xjlrztr6600000gp/T/azure_region-ss_enznj.log
[example_app] Test started. Log: less /var/folders/8f/56gzvwkd3n3293xjlrztr6600000gp/T/example_app-searsq5j.log
[minimal] Test started. Log: less /var/folders/8f/56gzvwkd3n3293xjlrztr6600000gp/T/minimal-d9h59mde.log
[gcp_image_id_dict_region] Test started. Log: less /var/folders/8f/56gzvwkd3n3293xjlrztr6600000gp/T/gcp_image_id_dict_region-p_mnkdtx.log
```

Michaelvll commented on examples/resnet_distributed_tf_app.py:
```diff
@@ -75,14 +75,19 @@ def run_fn(node_rank: int, ip_list: List[str]) -> Optional[str]:
 
 if __name__ == '__main__':
     cluster = None
+    cloud = None
```
It seems fine as we can allow it to be run on any cloud the user have when the example is used separately. We have made sure that the tests have `cloud` specified.

Michaelvll commented on sky/backends/backend_utils.py:
```diff
@@ -1678,6 +1678,9 @@ def _query_status_gcp(
     cluster: str,
     ray_config: Dict[str, Any],
 ) -> List[global_user_state.ClusterStatus]:
+    # Note: we use ":" for filtering labels for gcloud, as the latest gcloud (v393.0)
+    # fails to filter labels with "=".
+    # Reference: https://cloud.google.com/sdk/gcloud/reference/topic/filters
```
Yes, the problem does not happen for my older version as well, but when I try to run the smoke test on a newly launched GCP cluster. The problem occurs, which can cause the resource leakage.

If you upgrade the gcloud to v393.0, the following command will fail to fetch the info of an existing cluster.
```
gcloud compute instances list --filter="(labels.ray-cluster-name = mycluster)" --zones=us-west2-a --format="value(name)"
```

Michaelvll commented on sky/backends/cloud_vm_ray_backend.py:
```diff
@@ -2924,7 +2923,8 @@ def teardown_no_lock(self,
             elif ('TPU must be specified.' not in stderr and
-                  'SKYPILOT_ERROR_NO_NODES_LAUNCHED: ' not in stderr):
+                  'SKYPILOT_ERROR_NO_NODES_LAUNCHED: ' not in stderr and
+                  '(ResourceGroupNotFound)' not in stderr):
```
Good point! Added. Thanks!

Michaelvll replied:
Good point! Added. Thanks!

Michaelvll commented on sky/skylet/providers/gcp/node.py:
```diff
@@ -386,7 +386,7 @@ def _list_instances(self, label_filters: Optional[dict],
                 + " AND ".join(
                     [
-                        "(labels.{key} = {value})".format(key=key, value=value)
+                        "(labels.{key}:{value})".format(key=key, value=value)
```
The previous method seems not available now in the latest gcloud (v393.0), as described in

Michaelvll commented on sky/utils/tpu_utils.py:
```diff
@@ -78,8 +78,8 @@ def terminate_tpu_vm_cluster_cmd(cluster_name: str,
     query_cmd = (f'gcloud compute tpus tpu-vm list --filter='
-                 f'\\(labels.ray-cluster-name={cluster_name}\\) '
-                 f'--zone={zone} --format=value\\(name\\)')
+                 f'"(labels.ray-cluster-name:{cluster_name})" '
```
Yes, as mentioned above, the gcloud has deprecated the behavior of `=` in the latest version (v393) in my test.

Michaelvll commented on tests/conftest.py:
```diff
@@ -8,26 +8,75 @@
 # to mark a test as slow and to skip by default.
 # https://docs.pytest.org/en/latest/example/simple.html#control-skipping-of-tests-according-to-command-line-option
 
+clouds = ['aws', 'gcp', 'azure']
```
I would rather handle this manually, as the test writter should intentionally be aware of what tests are included.

Michaelvll commented on tests/conftest.py:
```diff
@@ -8,26 +8,75 @@
-        # --runslow given in cli: do not skip slow tests
-        return
     skip_slow = pytest.mark.skip(reason='need --runslow option to run')
+    skip_aws = pytest.mark.skip(reason='--no-aws is set, skip to run aws tests')
+    skip_gcp = pytest.mark.skip(reason='--no-gcp is set, skip to run gcp tests')
+    skip_azure = pytest.mark.skip(
+        reason='--no-azure is set, skip to run azure tests')
```
Done. Thanks!

Michaelvll commented on tests/conftest.py:
```diff
@@ -8,26 +8,75 @@
                      action='store_true',
                      default=False,
                      help='run slow tests')
+    for cloud in clouds:
+        parser.addoption(f'--{cloud}',
+                         action='store_true',
+                         default=True,
+                         help=f'use {cloud.upper()} for tests')
+        parser.addoption(f'--no-{cloud}',
+                         action='store_false',
+                         dest=cloud,
+                         help=f'do not use {cloud.upper()} for tests')
+    parser.addoption('--all-clouds',
+                     action='store_true',
+                     default=False,
+                     help='use all clouds for tests')
+    parser.addoption('--generic-cloud',
+                     type=str,
+                     default='gcp',
+                     choices=clouds,
+                     help='generic cloud to use for tests')
+    parser.addoption('--generic-spot-cloud',
+                     type=str,
+                     default='gcp',
+                     choices=clouds,
+                     help='generic spot cloud to use for tests')
```
Ahh, yes. It is expected, i.e. we disable the tests for aws by default with the `--disable-aws` in `pyproject.toml`. We can pass `pytest --enable-aws tests/test_smoke.py::test_custom_image`.

Added the following comment:
```
    # All markers (cluods/generic) except for `slow` are enabled by default.
    # To disable a cloud, use `--disable-<cloud>`. The generic tests are
    # cloud agnostic and can be run on any cloud, it is possible to specify
    # which cloud to use with `--generic-cloud` for generic tests and
    # `--generic-spot-cloud` for generic spot tests.
```

Michaelvll commented on tests/test_smoke.py:
```diff
@@ -134,33 +149,32 @@ def test_example_app():
 
 # ---------- A minimal task ----------
-def test_minimal():
+@pytest.mark.generic
```
Yes and yes, the AWS-only user should be able to trigger all generic tests with `--generic-cloud=aws`

Michaelvll commented on tests/test_smoke.py:
```diff
@@ -63,7 +75,10 @@ def _get_cluster_name() -> str:
+    test_name = caller_func_name.replace('_', '-').replace('test-', 't-')
+    if len(test_name) > 20:
+        assert len(test_name) < 46
+        test_name = 'test' + test_name[len('test')::2]
```
Truncating is a bit risky as that may cause two tests with name `test..._aws` and `test..._gcp` conflict with each other.

I changed the method to reduce the length with the following. Wdyt?
```
test_name = test_name[:20] + hashlib.md5(test_name.encode()).hexdigest()[:3]
```

Michaelvll commented on tests/test_smoke.py:
```diff
@@ -35,9 +35,21 @@
+# are running in parallel, the spot controller may be in INIT and
+# the spot queue command will return staled table.
+_SPOT_QUEUE_WAIT = ('s=$(sky spot queue); '
+                    'until [ `echo "$s" | grep INIT | wc -l` -eq 0 ]; '
```
Good point! Fixed. Thanks!

Michaelvll commented on tests/test_smoke.py:
```diff
@@ -271,31 +426,68 @@ def test_image_id_dict_with_zone():
             f'sky stop {name} -y',
             f'sky start {name} -y',
-            f'sky exec {name} examples/tests/test_yamls/no_conda_ami.yaml',
+            f'sky exec {name} tests/test_yamls/no_conda_ami.yaml',
```
Oops, it should be `examples/per_region_images.yaml`. Fixed. Thanks!

Michaelvll commented on pyproject.toml:
```diff
@@ -16,6 +16,7 @@ env = [
     "SKYPILOT_DEBUG=1",
     "SKYPILOT_DISABLE_USAGE_COLLECTION=1"
 ]
+addopts = "-s -n 16 -q --tb=short --disable-warnings --no-aws"
```
Done. Thanks!

Michaelvll commented:
> Thanks @Michaelvll! This is awesome. Read all files except test_smoke.
> 
> Some questions:
> 
> 1. I'm seeing some randomly generated names, why?
> 
> ```
> Terminating cluster testpiaei-itzn-2b289124-20...done.
> ...
> Terminating cluster testpsaejbmna-etr-2b289124-4f...done.
> ```

Good catch! I changed the name length reduction with truncate+hash as mentioned in https://github.com/skypilot-org/skypilot/pull/1587#discussion_r1070685581

> 2. `pytest tests/test_smoke.py -v --aws` was expecting it to run AWS-only tests, but seeing
> 
> ```
> ...
> [azure_region] Test started. Log: less /var/folders/8f/56gzvwkd3n3293xjlrztr6600000gp/T/azure_region-ss_enznj.log
> [example_app] Test started. Log: less /var/folders/8f/56gzvwkd3n3293xjlrztr6600000gp/T/example_app-searsq5j.log
> [minimal] Test started. Log: less /var/folders/8f/56gzvwkd3n3293xjlrztr6600000gp/T/minimal-d9h59mde.log
> [gcp_image_id_dict_region] Test started. Log: less /var/folders/8f/56gzvwkd3n3293xjlrztr6600000gp/T/gcp_image_id_dict_region-p_mnkdtx.log
> ```

By default, we enable the clouds, except for aws. `--aws` only enables the AWS but not affect other clouds.

concretevitamin commented:
> By default, we enable the clouds, except for aws. `--aws` only enables the AWS but not affect other clouds. To run aws only tests, it might need to be done by `pytest tests/test_smoke.py --disable-gcp --disable-azure --generic-cloud aws --generic-spot-cloud aws`
> 
> Do you think it would be more clear, if we disable all the clouds by default, and add `--enable-gcp` and `--enable-azure` in the `pyproject.toml`?

Thanks. I guess I was confused by remembering the "state" of what clouds are "enabled", and how that maps to what tests are run.

One proposal for discussion:
* no flags: run "generic" tests 
  - [ ] Q: what does generic mean? Those with `@pytest.mark.generic(_spot)` *plus* those without any markers?
* `--{aws, gcp, azure}`: with a flag `--<cloud>` set, run all generic tests plus `@pytest.mark.<cloud>`


This way, our default workflow would be 
```bash
pytest test_smoke.py  # Run all tests except AWS-specific ones.
pytest test_smoke.py --aws  # Make some AWS specific changes; test generic + AWS-specific.
```

----

Do we need both `--generic-cloud` and `--generic-spot-cloud`? Can we use only the former, to avoid confusion/burden to configure?

Michaelvll commented:
Thanks for the great comment! I changed the behavior for the options to be the one mentioned above:
```
pytest test_smoke.py  # Run all tests except AWS-specific ones.
pytest test_smoke.py --aws  # Make some AWS specific changes; test generic + AWS-specific.
pytest test_smoke.py --aws --gcp  # Make some AWS and GCP specific changes; test generic + AWS+GCP-specific.
```

I also removed the mark for generic and the `--generic-spot-cloud`. PTAL @concretevitamin : )

concretevitamin reviewed (approved):
This is awesome @Michaelvll! Thanks for significantly enhancing our testing infra - critical to ensuring correctness and coverage.

Michaelvll commented on CONTRIBUTING.md:
```diff
@@ -26,10 +26,19 @@ pip install -r requirements-dev.txt
+pytest tests/test_smoke.py --enable-aws
+
+# Change cloud for generic tests to aws
+pytest tests/test_smoke.py --generic-cloud aws
```
Good catch! Updated. Thanks!

Michaelvll commented on tests/conftest.py:
```diff
@@ -8,26 +8,76 @@
 # to mark a test as slow and to skip by default.
 # https://docs.pytest.org/en/latest/example/simple.html#control-skipping-of-tests-according-to-command-line-option
 
+clouds = ['aws', 'gcp', 'azure']
```
Done. Thanks!
